Fix boolean parsing in OpenVPNPlugin

Repsay opened the pull request:
Resolved an issue in the OpenVPNPlugin where the 'client-to-client' configuration option was incorrectly processed. Previously, the code attempted to use the value directly from the configuration, which led to errors when this value was an empty string. This fix ensures that an empty string for 'client-to-client' defaults to False, aligning with the expected boolean behavior of this configuration parameter.

pyrco commented:
Hi @Repsay, thanks for the bug fix!

Looking a bit further into the issue I think this issue is better fixed in the `_parse_config()` function. As I understand it OpenVPN allows options in its config file without any extra values to toggle them "on". The `duplicate_cn` option, used below the line you patched, for instance also works this way.

To have a general fix for these kind of options we could replace the line:
```
            value = value[0] if value else ""
```
in `_parse_config()` with:
```
            value = value[0] if value else True
```

Feel free to update your PR. Otherwise we can open a bug issue and will get to fixing it in due time.

Repsay commented:
Thank you for the suggestion to refine the issue in the `_parse_config()` function. I agree that a more general fix would be beneficial, particularly considering how OpenVPN processes options without extra values.

To address this, I've already updated the pull request. In addition to your suggestion, I've incorporated a dynamic way to identify boolean fields with the line `boolean_fields = OpenVPNServer.getfields("boolean") + OpenVPNClient.getfields("boolean")`. This enhancement should ensure that our parsing logic accurately reflects the nature of both OpenVPNServer and OpenVPNClient configurations.

Here's the revised section of the code:

```python
if any(key == boolean_field.name for boolean_field in boolean_fields):
    value = value[0] if value else True
else:
    value = value[0] if value else ""
    # This removes all text after the first comment
    value = CONFIG_COMMENT_SPLIT_REGEX.split(value, 1)[0].strip()
```

This update should offer a more robust and comprehensive solution to handle the OpenVPN configuration options correctly. Please review the changes in the updated PR, and let me know if you have any further suggestions or if additional modifications are required.

Repsay commented:
> @Repsay thank you for your contribution! As this is your first code contribution, please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following information:
> 
> ```
> @DissectBot agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > - (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > - (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> <details><summary>Contributor License Agreement</summary>
> <p>
> 
> ### Contribution License Agreement
> 
> This Contribution License Agreement ("_Agreement_") governs your Contribution(s) (as defined below) and conveys certain license rights to Fox-IT B.V. ("_Fox-IT_") for your Contribution(s) to Fox-IT"s open source Dissect project. This Agreement covers any and all Contributions that you ("_You_" or "_Your_"), now or in the future, Submit (as defined below) to this project. This Agreement is between Fox-IT B.V. and You and takes effect when you click an “I Accept” button, check box presented with these terms, otherwise accept these terms or, if earlier, when You Submit a Contribution. 
> 
> 1.	**Definitions.**
> 	"_Contribution_" means any original work of authorship, including any modifications or additions to an existing work, that is intentionally submitted by You to Fox-IT for inclusion in, or documentation of, any of the software products owned or managed by, or on behalf of, Fox-IT as part of the Project (the "_Work_").
> 	"_Project_" means any of the projects owned or managed by Fox-IT and offered under a license approved by the Open Source Initiative (www.opensource.org).
> 	"_Submit_" means any form of electronic, verbal, or written communication sent to Fox-IT or its representatives, including but not limited to communication on electronic mailing lists, source code control systems, and issue tracking systems that are managed by, or on behalf of, Fox-IT for the purpose of discussing and improving the Work, but excluding communication that is conspicuously marked or otherwise designated in writing by You as "_Not a Contribution_."
> 
> 2.	**Grant of Copyright License.** Subject to the terms and conditions of this Agreement, You hereby grant to Fox-IT and to recipients of software distributed by Fox-IT a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable copyright license to reproduce, prepare derivative works of, publicly display, publicly perform, sublicense, and distribute Your Contributions and such derivative works. 
> 
> 3.	**Grant of Patent License.** Subject to the terms and conditions of this Agreement, You hereby grant to Fox-IT and to recipients of software distributed by Fox-IT a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable (except as stated in this section) patent license to make, have made, use, maintain, offer to sell, sell, import, and otherwise transfer the Work, where such license applies only to those patent claims licensable by You that are necessarily infringed by Your Contribution(s) alone or by combination of Your Contribution(s) with the Work to which such Contribution(s) was submitted. If any entity institutes patent litigation against You or any other entity (including a cross-claim or counterclaim in a lawsuit) alleging that your Contribution, or the Work to which you have contributed, constitutes direct or contributory patent infringement, then any patent licenses granted to that entity under this Agreement for that Contribution or Work shall terminate as of the date such litigation is filed.
> 
> 4.	**Representations.** You represent that:
> 	- You are legally entitled to grant the above license. 
> 	- each of Your Contributions is Your original creation (see section 8 for submissions on behalf of others). 
> 	- Your Contribution submissions include complete details of any third-party license or other restriction (including, but not limited to, related patents and trademarks) of which you are personally aware and which are associated with any part of Your Contributions.
> 
> 5.	**Employer.** If Your Contribution is made in the course of Your work for an employer or Your employer has intellectual property rights in Your Submission by contract or applicable law, You must secure permission from Your employer to make the Contribution before signing this Agreement. In that case, the term "_You_" in this Agreement will refer to You and the employer collectively. If You change employers in the future and desire to Submit additional Contribution for the new employer, then You agree to sign a new Agreement and secure permission from the new employer before Submitting those Contributions.
> 
> 6.	**Support.** You are not expected to provide support for Your Contribution, unless You choose to do so. Any such support provided to the Project is provided free of charge. 
> 
> 7.	**Warranty.** Unless required by applicable law or agreed to in writing, You provide Your Contributions on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied, including, without limitation, any warranties or conditions of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A PARTICULAR PURPOSE.
> 
> 8.	**Third party material.** Should You wish to submit work that is not Your original creation, You may only submit it to Fox-IT separately from any Contribution, identifying the complete details of its source and of any license or other restriction (including, but not limited to, related patents, trademarks, and license agreements) of which You are personally aware, and conspicuously marking the work as "_Submitted on behalf of a third-party: [named here]_". 
> 
> 9.	**Notify.** You agree to notify Fox-IT of any facts or circumstances of which You become aware that would make the above representations inaccurate in any respect. 
> 
> 10.	**Governing law / competent court.** This Agreement is governed by the laws of the Netherlands. Any disputes that may arise are resolved by arbitration in accordance with the Arbitration Regulations of the Foundation for the Settlement of Automation Disputes (Stichting Geschillenoplossing Automatisering – SGOA – (www.sgoa.eu), this without prejudice to either party"s right to request preliminary relief in preliminary relief proceedings or arbitral preliminary relief proceedings. Arbitration proceedings take place in Amsterdam, or in any other place designated in the Arbitration Regulations. Arbitration shall take place in English. 
> </p>
> </details>

@DissectBot agree

pyrco commented on dissect/target/plugins/apps/vpn/openvpn.py:
```diff
@@ -166,14 +166,18 @@ def _parse_config(content: str) -> dict[str, Union[str, list[str]]]:
     """Parses Openvpn config  files"""
     lines = content.splitlines()
     res = {}
+    boolean_fields = OpenVPNServer.getfields("boolean") + OpenVPNClient.getfields("boolean")
```
I like the idea of special casing the known boolean fields. If you turn this into a set of field names, you can more easily check in the for loop and don't need an extra inner for loop anymore, e.g.:

```suggestion
    boolean_fields = OpenVPNServer.getfields("boolean") + OpenVPNClient.getfields("boolean")
    boolean_field_names = set(boolean_field.name for boolean_field in boolean_fields)
```

```
if key in boolean_field_names:
```

pyrco commented on dissect/target/plugins/apps/vpn/openvpn.py:
```diff
@@ -166,14 +166,18 @@ def _parse_config(content: str) -> dict[str, Union[str, list[str]]]:
+            else:
+                value = value[0] if value else ""
+                # This removes all text after the first comment
+                value = CONFIG_COMMENT_SPLIT_REGEX.split(value, 1)[0].strip()
```
I think we also need to deal with comments in configuration lines with boolean fields and split them off.

I was thinking along the lines:
```
key, _, value = line.partition(" ")
# This removes comments and whitespace around the value
value = CONFIG_COMMENT_SPLIT_REGEX.split(value, 1)[0].strip()
if not value and key in boolean_field_names:
    value = True
```

Repsay commented:
@pyrco, I've updated the code incorporating your recommendations. Please take a look and let me know if there's anything else you'd like adjusted.
